### ISSUE TITLE
Tests should always be built the same way, regardless of product build flavor

### ISSUE THREAD

**Issue by BruceForstall** [dotnet/coreclr#4724]
This is related to #4124

There are some tests, e.g. JIT\Regression\CLR-x86-JIT\V1-M09.5-PDC\b12053\b12053\b12053.exe, that are being built different between the Debug and Release build. In the Debug build, they appear to be built with /debug+ and the JIT generates debuggable code. The IL is different, e.g., in Debug there are additional NOP instructions, and the result of conditional expressions gets stored to locals.

These differences are bad for JIT testing, because the code we are testing against is different between a Debug JIT build and a Release JIT build.

**Comment by AndyAyersMS** [comment]
I wrote a parser to look at all the project files under the jit test tree. Roughly 1/3 of the tests will change behavior in Debug/Check vs Release.

For the cases that vary we need to decide whether to fix the tests as optimizable or not, or possibly duplicate them so we test both ways. 

Would also be g ood to cross check the result vs the "suffix conventions" (eg _r or _ro for optimizable code, _d or _do for debuggable code).

Not sure what a debug type of `Impl` means.

Some tests may also have source code conditional on `DEBUG` which can cause behavior to vary between build types. In some cases (eg the perf tests) this is intentional, but we should try and keep this to a minumum.

| DebugType | Condition | Count | WillTestVary | WillJitOptimize |
| --------- | --------- | ----- | ------------ | --------------- |
| Full      | No        | 1559  | No           | No              |
| PdbOnly   | No        | 1504  | No           | Yes             |
| None      | No        | 1006  | No           | Yes             |
| Impl      | No        |    2  | No           | ?               |
| (absent)  | No        | 1637  | Yes          | In Rel          |
| PdbOnly   | Yes       |  195  | Yes          | In Rel          |
| (blank)   | No        |  148  | Yes          | In Rel          |

**Comment by BruceForstall** [comment]
As I understand it, the suffix conventions correspond to C# compiler switches:

|suffix|switch|
|------|------|
| _r | /debug- /optimize- |
| _ro | /debug- /optimize+ |
| _d | /debug+ /optimize- |
| _do | /debug+ /optimize+ |

These were carried over from the JIT self-host (pre-GitHub) testbed. Some tests were (presumably) only ever built "ret" (I believe _ro) because we only ever used the RET build of the QA tree before. So not all tests had all flavors. If we were to continue this decision (and not reconsider all the tests that vary), we should just force "_ro" for all tests without a configuration.

@pgavlin worked on the transition to GitHub, so would have more to say.

**Comment by pgavlin** [comment]
> As I understand it, the suffix conventions correspond to C# compiler switches:

This is correct. These conventions were present in the original self-host testbed, and were used to indicate that a particular test would always be built with those switches regardless of the build flavor for the entire tree. These suffixes were also used for IL tests.

> @pgavlin worked on the transition to GitHub, so would have more to say.

The transition faithfully copied the compiler flags specified in a test's build description, so the tests that vary in the GitHub build would also have varied in the original test build. As @BruceForstall points out, we only ever used the RET build of the tests from the original testbed, so we were effectively using the `_ro` configuration for any test that did not specify otherwise.

I agree that using the flags for an `_ro` configuration for tests that currently vary would be equivalent to their behavior in the original testbed. We should consider whether or not there is any value in building these tests under other configurations as well (perhaps some subset--e.g. the bringup tests--should be built using all of the configurations listed above).

**Comment by BruceForstall** [comment]
I agree the CodeGenBringUpTests should be built in all flavors. (Note that in their pre-GitHub harness, they were -- they weren't part of the previous QA JIT self-host build system.)

**Comment by AndyAyersMS** [comment]
Looks like having `<DebugType></DebugType>` enables optimization too, so the above table was initially incorrect for the `(blank)` cases. Edited in place.

I've added some code that tries to verify the suffix convention vs the project file. Lots of data here and some  risk I've gotten it wrong, but the upshot is that around 300 suffixed tests have project settings that can disagree with their suffixes. Errors seem roughly equally distributed.

For instance:
```
JIT\Directed\coverage\importer\Desktop\badendfinally_il_r.ilproj -- no DebugType property, so test varies
JIT\Directed\shift\int16_d.csproj -- DebugType is blank --> release
JIT\jit64\hfa\main\testA\hfa_nd0A_d.csproj -- DebugType is pdbonly --> release
```

I'll work on extending the tool to rewrite the project files; this can be rule based so it can pick up whatever classification rules we deem appropriate.

**Comment by pgavlin** [comment]
> the upshot is that around 300 suffixed tests have project settings that can disagree with their suffixes

Rich and I noticed this sort of mismatch when doing the port, but IIRC we elected for fidelity to the original tests over agreement between suffixes and compiler flags.

**Comment by BruceForstall** [comment]
My guess is that people misunderstood the suffixes, and thus that the disagreements are more likely in the newer tests. E.g., I used to think "r" meant release==optimize.

Thus, it _might_ make more sense to fix the suffix than change the flags.

**Comment by AndyAyersMS** [comment]
For now I'll try and auto-fix all the files that don't have a suffix convention, aren't part of the bring up tests, and don't specify any DebugType.

The proposed fix will be to make them into "ro" tests but I won't change the actual file names.

**Comment by AndyAyersMS** [comment]
Fixed via #9712, save for the CodeGenBringUp tests, which require more work (#9722).
